kani-compiler/src/kani_middle/coercion.rs

```diff
@@ -0,0 +1,232 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! This module contains methods that help us process coercions.
+//! There are many types of coercions in rust, they are described in the
+//! [RFC 401 Coercions](https://rust-lang.github.io/rfcs/0401-coercions.html).
+//!
+//! The more complicated coercions are DST Coercions (aka Unsized Coercions). These coercions
+//! allow rust to create references to dynamically sized types (such as Traits and Slices) by
+//! casting concrete sized types. Unsized coercions can also be used to cast unsized to unsized
+//! types. These casts work not only on the top of references, but it also handle
+//! references inside structures, allowing the unsized coercions of smart pointers. This last
+//! piece was defined by
+//! [RFC 982 DST Coercion](https://rust-lang.github.io/rfcs/0982-dst-coercion.html).
```

Comment:
Are all unsized coercions constructing fat pointers of some kinds?

Also nit: can you be explicit about "this last piece". It's a stupid writing thing that bugs me, but I like less indirection in technical writing even if it means slightly more repetition.

Reply:
Yes. Done.